pkg/controller/logcollector/logcollector_controller.go

```diff
@@ -245,6 +247,67 @@ func (r *ReconcileLogCollector) Reconcile(request reconcile.Request) (reconcile.
 		}
 	}
 
+	if instance.Spec.AdditionalStores != nil {
+		if instance.Spec.AdditionalStores.Syslog != nil {
+			syslog := instance.Spec.AdditionalStores.Syslog
+
+			// Try to grab the ManagementClusterConnection CR (if present) because we need it for some
+			// validation with respect to Syslog.logTypes.
+			managementClusterConnection, mccErr := utils.GetManagementClusterConnection(ctx, r.client)
+
+			// In cases where Syslog configuration is present, but the logTypes field is
+			// empty or undefined, the default behavior is to enable all available logTypes.
+			if syslog.LogTypes == nil || len(syslog.LogTypes) == 0 {
+				// Set all available log types (except for v1.SyslogLogIDSEvents)
+				instance.Spec.AdditionalStores.Syslog.LogTypes = []v1.SyslogLogType{
+					v1.SyslogLogAuditEE,
+					v1.SyslogLogAuditKube,
+					v1.SyslogLogDNS,
+					v1.SyslogLogFlows,
+				}
+
+				// If this is not a managed cluster (i.e. no ManagementClusterConnection CR), then
+				// include v1.SyslogLogIDSEvents as well.
```

Comment:
@caseydavenport I can't remember if you were involved in discussing this behavior of changing the defaulting behavior based on ManagementClusterConnection CR existence.
I think I suggested this behavior but am second guessing it because there is a race condition of LogCollector CR being created and the ManagementClusterConnection CR.

Will it be a common occurrence to create the LogCollector CR and then later create the ManagementClusterConnection?

Reply:
Hm, I don't think I was involved.

I'm not a fan of this approach. Really, I think it's just not an API design that will fly.

 Like you say its relying on the order of events the user takes rather than the state that they provide, which breaks one of the more fundamental goals we're trying to provide in our API - for it to be fully declarative.

I think we just need to pick a default and stick with it, and then the user or our docs need to change it when it needs to be changed. The API is not intended to be zero-config for the user. Trying to remove the need for the user to ever configure the product while great in some limited capacity, when taken to the extreme is a "bad thing".

Reply:
What we _can_ do is have validation on this field (as well as on the ManagementClusterConnection CR) that makes sure this field doesn't include SyslogLogIDSEvents if this is a management cluster. 

I don't know the background... is it _really_ not allowed for this to be set on a managed cluster?

Reply:
> Hm, I don't think I was involved.

Yes, this wasn't discussed during the design phase. I unfortunately caught this late in the game (last week) and the discussion was mainly in [a Slack conversation](https://tigera.slack.com/archives/CJMB52RPG/p1602786223088700). 

Gist of it is that IDSEvent forwarding happens differently from the other log types because we're pulling the values directly out of the LogStorage (ES) which in an MCM setup is centralized. The other log types are always per cluster, so it behaves the same way regardless of the cluster (standalone/management/managed). 

Having this checking is more for the user's benefit as an extra warning so they know that for a managed cluster setting the IDSEvent option does nothing because it's not supported in a managed cluster. If you look at the logic below this comment thread, we don't add the IDSEvent during defaulting, unless the cluster is not a managed cluster. 



> I think we just need to pick a default and stick with it, and then the user or our docs need to change it when it needs to be changed. The API is not intended to be zero-config for the user. Trying to remove the need for the user to ever configure the product while great in some limited capacity, when taken to the extreme is a "bad thing".

Just to be clear the defaulting behaviour of logTypes isn't directly tied to this extra check. 

> Like you say its relying on the order of events the user takes rather than the state that they provide, which breaks one of the more fundamental goals we're trying to provide in our API - for it to be fully declarative.

My question is then, what is the declarative way to check this? The most certain way to determine whether this cluster happens to be managed is looking for the ManagementClusterConnection CR.

Reply:
> Just to be clear the defaulting behaviour of logTypes isn't directly tied to this extra check.

Really? I thought that this code was determining what entrires to put in the `logTypes` field when it is not defined, which is what I mean by defaulting :) Maybe I'm misunderstanding something, though? 

I think the correct way to do this is not to try to default the `logTypes` field based on whether or not a ManagementClusterConnection exists. 

Instead, always default to the same thing, and have _validation_ check that the provided configuration is consistent with itself. This validation can / should occur in both the logcollector and logstorage controllers.  

The problem with attempting to default based on the presence of another resource is that it takes something declarative - "here is my desired state" and turns it into something imperative - "here are the steps you need to take to get to my desired state". 

For the user, we don't want different behavior if they execute these commands:

```
kubectl create -f management-cluster-connection.yaml
kubeclt create -f log-storage.yaml
```

compared to if they run these commands:

```
kubeclt create -f log-storage.yaml
kubectl create -f management-cluster-connection.yaml
```

Reply:
An option we may want to consider:
Since `logTypes` is in a Syslog sub-struct under LogCollector, and not configured by default, maybe we should change it to a required field and not default it at all and only validate the field being set appropriately.
Though I guess that would be a problem on upgrades.  If we are defaulting the field to one of the sets (the one with or without IDSEvents) then we will be 'incorrectly' defaulting the logTypes for Managed or Management clusters.

Reply:
I am inclined to keep this as simple as possible. 

Making `logTypes` required and just forcing the user to set it makes a lot of sense to me (that's what I had prior to setting default values). 

> Though I guess that would be a problem on upgrades.

What would it be a problem on upgrades? I think it was only a problem when we were considering having default logic, but not explicitly declaring what the default values where in the CR. 

If we don't do any defaulting at all, the user has to explicitly set the field and then whatever they set carries over during upgrade (just like it would if we had set default values explicitly for them). 

Thoughts?

Reply:
> What would it be a problem on upgrades?
Currently if they have log forwarding configured, the user would see a change in behavior. Assuming there is no defaulting then the empty set is either invalid or will cause no logs to be forwarded which I would imagine will not be the behavior before the upgrade.

Reply:
Okay, I see you mean now Erik. 

I rather like the validation of having `logTypes` be required, because it puts the onus on the user to get their Syslog settings correct on their own. 

What about this setup: 
- Have the new `logTypes` field marked as required
    - This forces any new users that enable Syslog in v3.4 and beyond to set `logTypes` with whatever options they want
- For upgrades from previous versions, we keep a defaulting mechanism
    - The defaulting mechanism would work like what we have here already (set default options when it detects a LogCollector CR where the Syslog.LogTypes is empty/undefined and saves the change) 
   - We preserve the old behavior from before logTypes field **and** avoid this problem of checking whether it is a managed cluster or not, **by making the default behaviour only set the options to Audit, DNS and Flows (which is what it used to do)**
    - This means that the user has to explicitly turn on IDSEvents if they want it, which works well since it's new for v3.4 release 
    - If a user sets IDSEvent and this cluster happens to be in a managed cluster, we will also have a check for this that will communicate this isn't supported using the degraded status (whenever operator happens to detect a ManagementClusterConnection CR present)
        - This should be fine now because there won't be a race condition between defaulting (write to LogCollector) and checking for ManagementClusterConnection CR
    - Moving forward all new users who enable Syslog, will be forced to explicitly specify which log types they want because the field will be required and thus having explicit values will preserve expected behavior through into future releases 

Thoughts?

Reply:
Good summary. Seems reasonable to me.

Reply:
Okay, here is [the latest changes](https://github.com/tigera/operator/pull/941/commits/d931446f47ca10adbb9af8f7ff701bf7d4217c86), which follow the summary in [my previous comment](https://github.com/tigera/operator/pull/941/#discussion_r508866008) above.

I did some testing with a scenario where I had a LogCollector that did not have Syslog.LogTypes defined and then I upgraded to the latest operator and of the CRD. I can confirm that it correctly added the defaults (Audit, DNS, Flows). In any other case I am forced to set at least one value for LogTypes.